File: app/src/main/java/org/ametro/render/elements/DrawingElement.java
package org.ametro.render.elements;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.support.annotation.NonNull;


public abstract class DrawingElement implements Comparable<DrawingElement> {

    private Integer uid;

    private Rect boundingBox;

    private int priority;

    protected int layer;

    public Rect getBoundingBox() {
        return boundingBox;
    }

    public Integer getUid() {
        return uid;
    }

    public void setUid(Integer uid) {
        this.uid = uid;
    }

    public void setLayer(int layer) {
        this.layer = layer;
    }

    public abstract void draw(Canvas canvas);

    protected void setBoxAndPriority(Rect boundingBox, int priority) {
        this.boundingBox = boundingBox;
        this.priority = priority;
    }

    public int compareTo(@NonNull DrawingElement another) {
        int byLayer = this.layer - another.layer;
        if (byLayer != 0) {
            return byLayer;
        }

        int byPriority = this.priority - another.priority;
        if (byPriority != 0) {
            return byPriority;
        }

        return 0;
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/render/elements/DrawingElement.kt
package io.github.romangolovanov.apps.ametro.render.elements

import android.graphics.Canvas
import android.graphics.Rect

abstract class DrawingElement : Comparable<DrawingElement> {

    var uid: Int? = null

    private var boundingBox: Rect? = null

    private var priority: Int = 0

    @JvmField
    protected var layer: Int = 0

    fun getBoundingBox(): Rect? = boundingBox

    fun setLayer(layer: Int) {
        this.layer = layer
    }

    abstract fun draw(canvas: Canvas)

    protected fun setBoxAndPriority(boundingBox: Rect, priority: Int) {
        this.boundingBox = boundingBox
        this.priority = priority
    }

    override fun compareTo(other: DrawingElement): Int {
        val byLayer = this.layer - other.layer
        if (byLayer != 0) {
            return byLayer
        }
        val byPriority = this.priority - other.priority
        if (byPriority != 0) {
            return byPriority
        }
        return 0
    }
}
